### PR TITLE
#941 Add generic job information at the beginning of a job.

### PR DESCRIPTION
**Pull request by yruslan**
Also add a couple of additional metadata fields that might be helpful in certain circumstances.

### CODE REVIEW

File: utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala [hunk 1]
```diff
@@ -89,6 +119,19 @@ object PerformanceMetricTools {
     }
   }
 
+  /**
+    * Adds a Spark config key-value to Atum metadata if such key is present in Spark runtime config.
+    *
+    * @param optionPrefix A prefix for a job (e.g. "std", "conf", etc.)
+    * @param sparkKey A Spark configuration key
+    * @param atumKey An Atum metadata key
```
Comment (Zejnilovic) [inline utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala]: If you are aligning other comments, why not these?

File: utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala [hunk 2]
```diff
@@ -28,15 +28,62 @@ object PerformanceMetricTools {
...
+    addSparkConfig(optionPrefix, "spark.master", "spark_master")
+
+    Atum.setAdditionalInfo(s"${optionPrefix}_cmd_line_args" -> cmdLineArgs)
+    Atum.setAdditionalInfo(s"${optionPrefix}_input_dir" -> inputPath)
+    Atum.setAdditionalInfo(s"${optionPrefix}_output_dir" -> outputPath)
+    Atum.setAdditionalInfo(s"${optionPrefix}_input_dir_size" -> inputDirSize.toString)
+    Atum.setAdditionalInfo(s"${optionPrefix}_enceladus_version" -> ProjectMetadataTools.getEnceladusVersion)
+    Atum.setAdditionalInfo(s"${optionPrefix}_application_id" -> spark.sparkContext.applicationId)
+    Atum.setAdditionalInfo(s"${optionPrefix}_username" -> loginUserName)
+    Atum.setAdditionalInfo(s"${optionPrefix}_executors_num" -> s"$numberOfExecutors")
```
Comment (Zejnilovic) [inline utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala]: Isn't a number of executors also a `spakrConfig` option?

File: utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala [hunk 3]
```diff
@@ -28,15 +28,62 @@ object PerformanceMetricTools {
...
+    val inputDirSize = fsUtils.getDirectorySize(inputPath)
+
+    addSparkConfig(optionPrefix, "spark.driver.memory", "driver_memory")
+    addSparkConfig(optionPrefix, "spark.driver.cores", "driver_cores")
+    addSparkConfig(optionPrefix, "spark.driver.memoryOverhead", "driver_memory_overhead")
+    addSparkConfig(optionPrefix, "spark.executor.memory", "executor_memory")
+    addSparkConfig(optionPrefix, "spark.executor.cores", "executor_cores")
+    addSparkConfig(optionPrefix, "spark.executor.memoryOverhead", "executor_memory_overhead")
+    addSparkConfig(optionPrefix, "spark.submit.deployMode", "yarn_deploy_mode")
+    addSparkConfig(optionPrefix, "spark.master", "spark_master")
```
Comment (Zejnilovic) [inline utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala]: can we somehow pull the `--conf` data as well?

File: utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala (hunk 2 above)
Comment (yruslan) [reply utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala]: It's not always there. The way it gets the number of executors now is a "safer way".

Most of the listed options are absent at runtime unless explicitly specified in `spark-submit`. For instance, if the number of cores is not specified, this configuration key will be missing: `spark.executor.cores`. I assume that Spark will use the default value which is `1` for this setting.

File: utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala (hunk 3 above)
Comment (yruslan) [reply utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala]: Not quite following. Do you mean values such as Spline URL etc?

Comment (Zejnilovic) [reply utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala]: Yes, those. They can have things in that affect the job, right?

File: utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala (hunk 2 above)
Comment (Zejnilovic) [reply utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala]: Cool

File: utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala (hunk 3 above)
Comment (yruslan) [reply utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala]: There are not many config options available at runtime. I've included these that we already use and know they affect performance.
Specifically, JVM options I'm reluctant to include due to security reasons.

File: utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala [hunk 4]
```diff
@@ -28,15 +28,62 @@ object PerformanceMetricTools {
...
+
+    val fsUtils = new FileSystemVersionUtils(spark.sparkContext.hadoopConfiguration)
+    // Directory sizes and size ratio
+    val inputDirSize = fsUtils.getDirectorySize(inputPath)
+
+    addSparkConfig(optionPrefix, "spark.driver.memory", "driver_memory")
+    addSparkConfig(optionPrefix, "spark.driver.cores", "driver_cores")
+    addSparkConfig(optionPrefix, "spark.driver.memoryOverhead", "driver_memory_overhead")
+    addSparkConfig(optionPrefix, "spark.executor.memory", "executor_memory")
+    addSparkConfig(optionPrefix, "spark.executor.cores", "executor_cores")
```
Comment (DzMakatun) [inline utils/src/main/scala/za/co/absa/enceladus/utils/performance/PerformanceMetricTools.scala]: Very happy to see the executor cores included! 👍

**Comment by yruslan** [comment]
### For release notes:
* Run metadata is added at the beginning of a Spark job so if the job fails users have additional information for debugging
* Added additional Spark configuration parameters to run's metadata.

**Pull request merged**